Cover Pixel

kamilahjae opened the pull request:
Creates cover pixel functionality to
blend colors with alpha values greater
than zero. Also expands RGB values to
RGBA.

kamilahjae committed:
Cover Pixel

Creates cover pixel functionality to
blend colors with alpha values greater
than zero.

kamilahjae committed:
Cover Pixel

Creates cover pixel functionality to
blend colors with alpha values greater
than zero.

AKimZ commented on render.h:
```diff
@@ -8,11 +8,17 @@
+// much cheaper to calculate
+// for values between 0 and 65534
+#define div255( num ) ( ( num + ( num >> 8 ) ) >> 8 )
+#define lerp255( background, foreground, a ) ( ( uint8_t )div255( ( foreground * a + background * ( 255-a ) ) ) )
```
lerp?

AKimZ commented on draw.c:
```diff
@@ -8,12 +8,12 @@
             fclose(out);
 }
 
-/* Make the picture, store it, and o home */
```
:stuck_out_tongue_closed_eyes:

AKimZ commented on draw.c:
```diff
@@ -8,12 +8,12 @@
 {
     FILE *out = fopen (filename, "w");
-        fprintf(out, "P6 %d %d %d\n", fb->width, fb->height, 255 );
+        fprintf(out, "P8 %d %d %d\n", fb->width, fb->height, 255 );
```
Is this where you said that write ppm wasn't accepted 8 bit data?

kamilahjae replied:
@AKimZ: that was an incorrect assumption on my part, the P6 is a "magic number" to specify the file type  ¯_(ツ)_/¯

AKimZ commented:
I'd be interested in understanding more about how this works. Maybe you can walk me through it when we're less feverish? (We can merge it in if it's stable and compiles though)

kamilahjae committed:
Create a LuaJit Wrapper for C API

Creates a luaJit wrapper to use C
library inside of the lua scripting
language.

kamilahjae committed:
Merge Master into CoverPixel Branch

kamilahjae committed:
Pixel and Frame Buffer Optimizations

Optimizes Pixel bits with #define for
speed and refactors frameBuffer to
include valid "ownsData" conditional.
Creates new file for colors. Removes
swap macro from header file. Refactors
write PPM function to handle RGBA color
values in draw.c

kamilahjae committed:
Merge branch 'master' into coverPixel

Conflicts:
	draw_ffi.lua

kamilahjae commented:
@AKimZ:  lerp stands for linear interpolation: https://unity3d.com/learn/tutorials/modules/beginner/scripting/linear-interpolation. In the case for the color values, I'm finding the value between an r, g, or b color value and the alpha value. Like interpolating between the values of 5 and 7 by 50% to get 6.

kamilahjae commented:
@AKimZ This is working and up-to-date with master! :boom:  Currently using the coverPixel function on the horizontal line function.
